zetaclient/orchestrator/orchestrator.go

```diff
@@ -289,122 +289,113 @@ func (oc *Orchestrator) runScheduler(ctx context.Context) error {
 		return err
 	}
 
-	observeTicker := time.NewTicker(3 * time.Second)
-	var lastBlockNum int64
+	newBlockChan, err := oc.zetacoreClient.NewBlockSubscriber(ctx)
+	if err != nil {
+		return err
+	}
+
 	for {
 		select {
 		case <-oc.stop:
 			oc.logger.Warn().Msg("runScheduler: stopped")
 			return nil
-		case <-observeTicker.C:
-			{
-				bn, err := oc.zetacoreClient.GetBlockHeight(ctx)
+		case <-time.After(time.Second * 10):
+			// the subscription should automatically reconnect after zetacore
+			// restart, but we should log this just in case that logic is not
+			// working
+			oc.logger.Warn().Msg("runScheduler: no blocks after 10 seconds")
+		case newBlock := <-newBlockChan:
```

Comment:
There is a slight difference I noticed when compared to non-subscription model. So can `newBlockChan` hold multiple items (blocks) inside? For example, if it contains `[block100, block101, block102]` somehow (due to delay), do we still action on old blocks `block100, block101`? Th non-subscription model only deal with whatever block latest.

Reply:
> do we still action on old blocks

We are not actually doing work in response to block, that comes from the `GetPendingCctxsWithinRateLimit`. There is already logic which will prevent outbound from being processed again during the next tick.

I don't think this is a huge issue but I could add another check to just ignore block notifications if the block time is older than 10 seconds or something like that.

Reply:
yeah, ignoring block time older than 10 seconds or maybe just set the buffered channel size as 1 or 2, like
`blockEventChan := make(chan cometbfttypes.EventDataNewBlock, 2)`. You could do whatever convenient to you, just in case the channel gets blocks piled up.